sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisIO.java

```diff
@@ -522,4 +651,393 @@ public Read withMaxCapacityPerShard(Integer maxCapacity) {
+
+    /**
+     * Specify how to partition records among all stream shards (required).
+     *
+     * <p>The partitioner is critical to distribute new records among all stream shards.
+     */
+    public Write<T> withPartitioner(KinesisPartitioner<T> partitioner) {
+      return builder().partitioner(partitioner).build();
+    }
+
+    /** Specify how to serialize records to bytes on the stream (required). */
+    public Write<T> withSerializer(SerializableFunction<T, byte[]> serializer) {
+      return builder().serializer(serializer).build();
+    }
+
+    /**
+     * Specify a custom {@link KinesisAsyncClientProvider} providing a {@link KinesisAsyncClient} to
+     * communication with Kinesis.
+     *
+     * <p>When sufficient use {@link Write#withKinesisClientProvider(AwsCredentialsProvider,
+     * Region)}.
+     */
+    public Write<T> withKinesisClientProvider(KinesisAsyncClientProvider clientProvider) {
+      return builder().kinesisClientProvider(clientProvider).build();
+    }
+
+    /**
+     * Specify the {@link AwsCredentialsProvider} and AWS region to be used to communicate with
+     * Kinesis.
+     */
+    public Write<T> withKinesisClientProvider(
+        AwsCredentialsProvider credentialsProvider, Region region) {
+      return withKinesisClientProvider(new BasicKinesisProvider(credentialsProvider, region, null));
+    }
+
+    /**
+     * Specify the {@link AwsCredentialsProvider} and AWS region to be used to communicate with
+     * Kinesis. *
+     *
+     * <p>The {@code serviceEndpoint} sets an alternative service host. This is useful to execute
+     * the tests with a kinesis service emulator.
+     */
+    Write<T> withKinesisClientProvider(
+        AwsCredentialsProvider credentialsProvider, Region region, String endpoint) {
+      return withKinesisClientProvider(
+          new BasicKinesisProvider(credentialsProvider, region, endpoint));
+    }
+
+    /** Configure the AWS SDK retry behavior of the {@link KinesisAsyncClient}. */
+    public Write<T> withRetryConfiguration(RetryConfiguration retryConfiguration) {
+      return builder().retryConfiguration(retryConfiguration).build();
+    }
+
+    public Write<T> validated() {
+      checkArgument(!isEmpty(streamName()), "streamName cannot be empty");
+      checkArgument(partitioner() != null, "partitioner cannot be null");
+      checkArgument(serializer() != null, "serializer cannot be null");
+      checkArgument(kinesisClientProvider() != null, "kinesisClientProvider cannot be null");
+      checkArgument(concurrentRequests() > 0, "concurrentRequests must be > 0");
+      checkArgument(
+          batchMaxRecords() > 0 && batchMaxRecords() <= MAX_RECORDS_PER_REQUEST,
+          "batchMaxRecords must be in [1,%s]",
+          MAX_RECORDS_PER_REQUEST);
+      checkArgument(
+          batchMaxBytes() > 0 && batchMaxBytes() <= MAX_BYTES_PER_REQUEST,
+          "batchMaxBytes must be in [1,%s]",
+          MAX_BYTES_PER_REQUEST);
+      checkArgument(
```

Comment:
Better to check it in `aggregatedRecordsMaxBytes()`

Reply:
ping

Reply:
sorry, fixed already